File: zh-Hant/codes/cpp/chapter_backtracking/permutations_ii.cpp
/**
 * File: permutations_ii.cpp
 * Created Time: 2023-04-24
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* 回溯演算法：全排列 II */
void backtrack(vector<int> &state, const vector<int> &choices, vector<bool> &selected, vector<vector<int>> &res) {
    // 當狀態長度等於元素數量時，記錄解
    if (state.size() == choices.size()) {
        res.push_back(state);
        return;
    }
    // 走訪所有選擇
    unordered_set<int> duplicated;
    for (int i = 0; i < choices.size(); i++) {
        int choice = choices[i];
        // 剪枝：不允許重複選擇元素 且 不允許重複選擇相等元素
        if (!selected[i] && duplicated.find(choice) == duplicated.end()) {
            // 嘗試：做出選擇，更新狀態
            duplicated.emplace(choice); // 記錄選擇過的元素值
            selected[i] = true;
            state.push_back(choice);
            // 進行下一輪選擇
            backtrack(state, choices, selected, res);
            // 回退：撤銷選擇，恢復到之前的狀態
            selected[i] = false;
            state.pop_back();
        }
    }
}

/* 全排列 II */
vector<vector<int>> permutationsII(vector<int> nums) {
    vector<int> state;
    vector<bool> selected(nums.size(), false);
    vector<vector<int>> res;
    backtrack(state, nums, selected, res);
    return res;
}

/* Driver Code */
int main() {
    vector<int> nums = {1, 1, 2};

    vector<vector<int>> res = permutationsII(nums);

    cout << "輸入陣列 nums = ";
    printVector(nums);
    cout << "所有排列 res = ";
    printVectorMatrix(res);

    return 0;
}

File: ja/codes/rust/chapter_backtracking/permutations_ii.rs
/*
 * File: permutations_ii.rs
 * Created Time: 2023-07-15
 * Author: codingonion (coderonion@gmail.com)
 */

use std::collections::HashSet;

/* バックトラッキング：順列 II */
fn backtrack(mut state: Vec<i32>, choices: &[i32], selected: &mut [bool], res: &mut Vec<Vec<i32>>) {
    // 状態の長さが要素数に等しければ、解を記録
    if state.len() == choices.len() {
        res.push(state);
        return;
    }
    // すべての選択肢を走査
    let mut duplicated = HashSet::<i32>::new();
    for i in 0..choices.len() {
        let choice = choices[i];
        // 枝刈り：要素の重複選択を許可せず、同値要素の重複選択も許可しない
        if !selected[i] && !duplicated.contains(&choice) {
            // 試行: 選択を行い、状態を更新
            duplicated.insert(choice); // 選択済みの要素値を記録
            selected[i] = true;
            state.push(choice);
            // 次の選択へ進む
            backtrack(state.clone(), choices, selected, res);
            // バックトラック：選択を取り消し、前の状態に戻す
            selected[i] = false;
            state.pop();
        }
    }
}

/* 全順列 II */
fn permutations_ii(nums: &mut [i32]) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    backtrack(Vec::new(), nums, &mut vec![false; nums.len()], &mut res);
    res
}

/* Driver Code */
pub fn main() {
    let mut nums = [1, 2, 2];

    let res = permutations_ii(&mut nums);

    println!("入力配列 nums = {:?}", &nums);
    println!("すべての順列 res = {:?}", &res);
}
